### PR TITLE
feat: BeneficialOwners step update

### PR DESCRIPTION
**Pull request by MrMuzyk**
### Details

- Updated BeneficialOwners step
- added an option to exit the process at any point
- fixed issues that came up after rebase with main

### Fixed Issues

$ https://github.com/Expensify/App/issues/30435
PROPOSAL:
not applicable here

### Tests

- [ ] Verify that no errors appear in the JS console
- [ ] get through VBBA process up to Beneficial Owners step
- [ ] check if you can perform all possible scenarios

### QA Steps

- [ ] Verify that no errors appear in the JS console
- [ ] get through VBBA process up to Beneficial Owners step
- [ ] check if you can perform all possible scenarios

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/23487612/32759bfd-bea0-44ce-a718-11e5e8cdb2bb

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/23487612/cc82cdaf-33db-432e-b1bc-c962521c6fcd

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/23487612/d26863bb-961d-4eb1-8065-51a21fe1c7b0

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/23487612/e008117d-416d-4dcc-820d-32e97ab792a7

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/23487612/2250753c-2b9b-4219-ac7e-dd0cb3f8967d

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/23487612/55181a55-ea58-4f72-a793-f769047186e1

</details>

### CODE REVIEW

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.js [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    /** Changes variable responsible for displaying step 4 or 5 */
+    setIsBeneficialOwnerInfoSet: PropTypes.func.isRequired,
+};
+
+const defaultProps = {
+    reimbursementAccount: ReimbursementAccountProps.reimbursementAccountDefaultProps,
+    reimbursementAccountDraft: {},
+};
+
+const bodyContent = [LegalNameUBO, DateOfBirthUBO, SocialSecurityNumberUBO, AddressUBO, ConfirmationUBO];
```
Comment (rezkiy37) [inline src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.js:53]: Let's use UPPER_CASE for const to standardise them. Same for next const.
```suggestion
const BODY_CONTENT = [LegalNameUBO, DateOfBirthUBO, SocialSecurityNumberUBO, AddressUBO, ConfirmationUBO];
```

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.js [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+const substep = CONST.BANK_ACCOUNT.BENEFICIAL_OWNER_INFO_STEP.SUBSTEP;
+const MAX_NUMBER_OF_UBOS = 4;
+
+function BeneficialOwnerInfo({reimbursementAccount, reimbursementAccountDraft, onBackButtonPress, onCloseButtonPress, setIsBeneficialOwnerInfoSet}) {
+    const {translate} = useLocalize();
+    const companyName = getDefaultValueForReimbursementAccountField(reimbursementAccount, 'companyName', '');
+    const values = useMemo(() => getSubstepValues(beneficialOwnerInfoStepKeys, reimbursementAccountDraft, reimbursementAccount), [reimbursementAccount, reimbursementAccountDraft]);
+    const defaultBeneficialOwnerKeys = lodashGet(reimbursementAccountDraft, beneficialOwnerKeysKey, []);
+
+    // TODO we're only reading beneficialOwnerKeys from draft values because there is not option to remove UBO
```
Comment (rezkiy37) [inline src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.js:65]: Should this TODO be merged?

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.js [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        // Reset Beneficial Owner Details Form to first substep
+        resetScreenIndex();
+        setCurrentUBOSubstep(substep.UBO_DETAILS_FORM);
+    };
+
+    const handleNextUBOSubstep = (value) => {
+        if (currentUBOSubstep === substep.IS_USER_UBO) {
+            setIsUserUBO(value);
+
+            // user is an owner but there are 4 other owners already added, so we remove last one
```
Comment (rezkiy37) [inline src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.js:146]: Same for other comments.
```suggestion
            // User is an owner but there are 4 other owners already added, so we remove last one
```

**Review by rezkiy37 — CHANGES_REQUESTED** [review]
Left comments, after LGTM.

**Comment by MrMuzyk** [comment]
@shubham1206agra Ready for review

**Comment by shubham1206agra** [comment]
@MrMuzyk Can you quickly resync your branch?

**Comment by shubham1206agra** [comment]
It crashes now

**Comment by MrMuzyk** [comment]
> It crashes now

what exactly?

**Comment by shubham1206agra** [comment]
Bank Account

**Comment by MrMuzyk** [comment]
Checking

**Comment by shubham1206agra** [comment]
This is happening due to we are using styles instead of useThemeStyles

**Comment by shubham1206agra** [comment]
There are way too many import errors

**Comment by shubham1206agra** [comment]
Screenshot 2023-12-14 at 5 37 12 PM
I am now getting this error on step 2 starting

**Comment by shubham1206agra** [comment]
Screenshot 2023-12-14 at 6 56 13 PM

This is defaulted to No

**Comment by shubham1206agra** [comment]
Screenshot 2023-12-14 at 7 00 06 PM

Wrong placeholder

**Comment by MrMuzyk** [comment]
Its a radio button, it needs to have a value. Default value is false

> Screenshot 2023-12-14 at 6 56 13 PM
> This is defaulted to previous option selected for some reason

I dont think so, here is the code -     `const [isAnyoneElseUBO, setIsAnyoneElseUBO] = useState(values[BENEFICIAL_OWNER_INFO_STEP_KEYS.BENEFICIAL_OWNERS].length > 0);`

**Comment by MrMuzyk** [comment]
> Screenshot 2023-12-14 at 7 00 06 PM
> Wrong placeholder

can you please leave the comments in the code? Will be easier for me to track all the things that require fixing

**Comment by MrMuzyk** [comment]
If something isnt related to specific line of code post a video or something obviously

**Comment by shubham1206agra** [comment]
Screenshot 2023-12-14 at 7 05 28 PM
Maybe text can be `Do any other individual` whenever this question is asked again

**Comment by shubham1206agra** [comment]
https://github.com/Expensify/App/assets/58412969/d9f452e5-13f4-4ba6-98f0-7b1540c00415

See this. Getting different behaviour now

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/substeps/BeneficialOwnerDetailsFormSubsteps/AddressUBO.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            style={[styles.mh5, styles.flexGrow1]}
+        >
+            <Text style={[styles.textHeadline]}>{translate('beneficialOwnerInfoStep.enterTheOwnersAddress')}</Text>
+            <Text>{translate('common.noPO')}</Text>
+            <AddressForm
+                inputKeys={INPUT_KEYS}
+                shouldSaveDraft
+                translate={translate}
+                defaultValues={defaultValues}
+                streetTranslationKey="common.companyAddress"
```
Comment (shubham1206agra) [inline src/pages/ReimbursementAccount/BeneficialOwnerInfo/substeps/BeneficialOwnerDetailsFormSubsteps/AddressUBO.js:83]: This placeholder is wrong

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.js [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+                <BeneficialOwnerCheckUBO
+                    title={`${translate('beneficialOwnerInfoStep.doYouOwn25percent')} ${companyName}?`}
+                    defaultValue={isUserUBO}
+                    onSelectedValue={handleNextUBOSubstep}
+                />
+            )}
+
+            {currentUBOSubstep === SUBSTEP.IS_ANYONE_ELSE_UBO && (
+                <BeneficialOwnerCheckUBO
+                    title={`${translate('beneficialOwnerInfoStep.doAnyIndividualOwn25percent')} ${companyName}?`}
```
Comment (shubham1206agra) [inline src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.js:254]: Maybe text can be `Do any other individual` whenever this question is asked again

Comment (MrMuzyk) [reply src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.js:254]: Use designs as a reference when it comes to texts. I dont have the authority to change the texts we display

**Comment by MrMuzyk** [comment]
> Screen.Recording.2023-12-14.at.7.07.56.PM.mov 
> See this. Getting different behaviour now

I see, know where the issue is. Fixing it

**Comment by MrMuzyk** [comment]
@shubham1206agra Placeholder and default value should be fixed

**Comment by shubham1206agra** [comment]
https://github.com/Expensify/App/assets/58412969/01db475e-948d-40ac-9593-e1a7af0c4b6c

I am getting this warning here

**Comment by shubham1206agra** [comment]
Screenshot 2023-12-14 at 9 35 58 PM

Android is not compiling for me currently

**Comment by shubham1206agra** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/58412969/76e3df92-a84b-46dc-aee2-6268a228f76f

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/58412969/4690d7b8-fb1d-40f5-8b80-f23b3805cb26

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/58412969/936b7bdd-c1a5-469a-84e6-464e6deef220

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/58412969/f0f44da8-e1fa-4728-9538-19238f501e7c

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/58412969/b4cc19ed-8abe-45f0-b8f8-b18cef1fe7c9

</details>

**Comment by shubham1206agra** [comment]
@MrMuzyk Can you address this https://github.com/Expensify/App/pull/32709#issuecomment-1856063371 so I can approve this PR?

**Comment by shubham1206agra** [comment]
Still waiting on this comment

**Comment by MrMuzyk** [comment]
@shubham1206agra Please check again now

**Review by shubham1206agra — APPROVED** [review]

**Pull request merged**